public/jira/raw/rollback/rollback.ts

```diff
@@ -0,0 +1,311 @@
+  }
+}
+
+type FieldRollbackFunction = (
+  lastReturnValue: unknown,
+  change: Change,
+  fieldName: string,
+  data: RollbackLookupData
+) => { [key: string]: unknown };
+
+type FieldRollbackFunctionMap = { [fieldName: string]: FieldRollbackFunction };
+
+export const fields: FieldRollbackFunctionMap = {
+  // from will look like "1619, 1647"
+  // we need to update `lastReturnValue` to have
+  // only the right sprints
+  Sprint: function (lastReturnValue, change, fieldName, { sprints }) {
+    const sprintNumbers = getSprintNumbers(change.from || "");
+    const sprintNames = getSprintNames(change.fromString || "");
+
+    if (sprintNumbers === null) {
+      return { [fieldName]: null };
+    } else {
+      return {
+        [fieldName]: sprintNumbers
+          .map((number, i) => {
+            // REMOVE IN PROD
+            if (sprints.ids.has(number)) {
+              return sprints.ids.get(number);
+            } else if (sprints.names.has(sprintNames![i])) {
+              return sprints.names.get(sprintNames![i]);
+            } else {
+              // TODO: change to async so we can go request all of these
+              console.warn("Can't find sprint ", number, sprintNames![i]);
+              return undefined;
+            }
+          })
+          .filter((x): x is Sprint => x !== undefined),
+      };
+    }
+  },
+  "Fix versions": function (lastReturnValue, change, fieldName, { versions }) {
+    if (change.from) {
+      if (versions.ids.has(change.from)) {
+        return { [fieldName]: versions.ids.get(change.from) };
+      } else if (versions.names.has(change.fromString || "")) {
+        return { [fieldName]: versions.names.get(change.fromString || "") };
+      } else {
+        console.warn(
+          "Can't find release version ",
+          change.from,
+          change.fromString
+        );
+        return { [fieldName]: lastReturnValue };
+      }
+    } else {
+      return { [fieldName]: [] };
+    }
+  },
+  // Parent Link, Epic Link,
+  IssueParentAssociation: function (lastReturnValue, change) {
+    return { Parent: { key: change.fromString || "", id: change.from || "" } };
+  },
+  "Parent Link": function (lastReturnValue: unknown, change: Change) {
+    return { Parent: { key: change.fromString || "" } };
+  },
+  "Epic Link": function (lastReturnValue: unknown, change: Change) {
+    return { Parent: { key: change.fromString || "" } };
+  },
+  Status: function (lastReturnValue, change, fieldName, { statuses }) {
+    if (statuses.ids.has(change.from || "")) {
+      return { [fieldName]: statuses.ids.get(change.from || "") };
+    } else if (statuses.names.has(change.fromString || "")) {
+      return { [fieldName]: statuses.names.get(change.fromString || "") };
+    } else {
+      console.warn("Can't find status", change.from, change.fromString);
+      return { [fieldName]: { name: change.fromString || "" } };
+    }
+  },
+};
+
+const fieldAlias: { [key: string]: string } = {
```

Comment:
interesting why `{[key: string]: string}` ? could we just leave it to inference?

Reply:
Yup the return type should be obvious, no need to explicitly add it

Reply:
Its usage below will complain because it sees the type as
```
{
    duedate: string;
    status: string;
    labels: string;
    issuetype: string;
    "Fix Version": string;
}
```
When it access it like `fieldAlias[field]` 
`Element implicitly has an 
 type because expression of type 
 can't be used to index type:`
 
 so either I can cast the usage below to `fieldAlias[field as keyof typeof fieldAlias] `
 or explicitly force this type